companion/src/storage/categorized.cpp

```diff
@@ -136,20 +143,33 @@ bool CategorizedStorageFormat::write(const RadioData & radioData)
     if (!writeFile(modelData, modelFilename)) {
       return false;
     }
-    if (getCurrentFirmware()->getCapability(HasModelCategories)) {
-      int categoryIndex = model.category;
-      if (currentCategoryIndex != categoryIndex) {
-        modelsList.append(QString().sprintf("[%s]\n", radioData.categories[model.category].name));
-        currentCategoryIndex = categoryIndex;
-      }
-    }
-    if (IS_HORUS(getCurrentBoard())) {
-      modelsList.append(QString("%1\n").arg(model.filename));
+
+    currentCategoryIndex = getCurrentFirmware()->getCapability(HasModelCategories) ?
+                  model.category : -1;
+    sortedModels[currentCategoryIndex].push_back(m);
+  }
+
+  for (int c= -1; c < (int)numCategories; c++) {
+
+    if (c > -1 && getCurrentFirmware()->getCapability(HasModelCategories)) {
+      modelsList.append(QString().sprintf("[%s]\n", radioData.categories[c].name));
     }
-    else {
-      // use format with model number and file name
-      // this is needed, because this kind of radios can have unused model slots
-      modelsList.append(QString("%1 %2\n").arg(i).arg(model.filename));
+
+    numModels = sortedModels[c].size();
+    for (unsigned int m=0; m<numModels; m++) {
+      const ModelData & model = radioData.models[sortedModels[c][m]];
+      QString line;
+      if (IS_HORUS(getCurrentBoard())) {
+        line = QString("%1\n").arg(model.filename);
+      }
+      else {
+        // use format with model number and file name
+        // this is needed, because this kind of radios can have unused model slots
+        // NOTE: The model number will only be unique within a category
+        line = QString("%1\n").arg(model.filename);
```

Comment:
This won't work on radios that have no categories. Such radios need a different line format with `<model number> <model name>` because of #4295.

So for such radios, be sure to keep the old code (in this case there is no need for map<> because you have no categories.

For radios that have categories your code looks good.